src/node_http2.cc

```diff
@@ -0,0 +1,1326 @@
+      OneByteString(env->isolate(), nghttp2_strerror(val)));
+}
+
+// Serializes the settings object into a Buffer instance that
+// would be suitable, for instance, for creating the Base64
+// output for an HTTP2-Settings header field.
+void PackSettings(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  HandleScope scope(env->isolate());
+
+  std::vector<nghttp2_settings_entry> entries;
+  entries.reserve(6);
+
+  uint32_t* const buffer = env->http2_settings_buffer();
+  uint32_t flags = buffer[IDX_SETTINGS_COUNT];
+
+  if ((flags & (1 << IDX_SETTINGS_HEADER_TABLE_SIZE)) ==
+      (1 << IDX_SETTINGS_HEADER_TABLE_SIZE)) {
+    DEBUG_HTTP2("Setting header table size: %d\n",
+                buffer[IDX_SETTINGS_HEADER_TABLE_SIZE]);
+    entries.push_back({NGHTTP2_SETTINGS_HEADER_TABLE_SIZE,
+                       buffer[IDX_SETTINGS_HEADER_TABLE_SIZE]});
+  }
+
+  if ((flags & (1 << IDX_SETTINGS_MAX_CONCURRENT_STREAMS)) ==
+      (1 << IDX_SETTINGS_MAX_CONCURRENT_STREAMS)) {
+    DEBUG_HTTP2("Setting max concurrent streams: %d\n",
+                buffer[IDX_SETTINGS_MAX_CONCURRENT_STREAMS]);
+    entries.push_back({NGHTTP2_SETTINGS_MAX_CONCURRENT_STREAMS,
+                       buffer[IDX_SETTINGS_MAX_CONCURRENT_STREAMS]});
+  }
+
+  if ((flags & (1 << IDX_SETTINGS_MAX_FRAME_SIZE)) ==
+      (1 << IDX_SETTINGS_MAX_FRAME_SIZE)) {
+    DEBUG_HTTP2("Setting max frame size: %d\n",
+                buffer[IDX_SETTINGS_MAX_FRAME_SIZE]);
+    entries.push_back({NGHTTP2_SETTINGS_MAX_FRAME_SIZE,
+                       buffer[IDX_SETTINGS_MAX_FRAME_SIZE]});
+  }
+
+  if ((flags & (1 << IDX_SETTINGS_INITIAL_WINDOW_SIZE)) ==
+      (1 << IDX_SETTINGS_INITIAL_WINDOW_SIZE)) {
+    DEBUG_HTTP2("Setting initial window size: %d\n",
+                buffer[IDX_SETTINGS_INITIAL_WINDOW_SIZE]);
+    entries.push_back({NGHTTP2_SETTINGS_INITIAL_WINDOW_SIZE,
+                       buffer[IDX_SETTINGS_INITIAL_WINDOW_SIZE]});
+  }
+
+  if ((flags & (1 << IDX_SETTINGS_MAX_HEADER_LIST_SIZE)) ==
+      (1 << IDX_SETTINGS_MAX_HEADER_LIST_SIZE)) {
+    DEBUG_HTTP2("Setting max header list size: %d\n",
+                buffer[IDX_SETTINGS_MAX_HEADER_LIST_SIZE]);
+    entries.push_back({NGHTTP2_SETTINGS_MAX_HEADER_LIST_SIZE,
+                       buffer[IDX_SETTINGS_MAX_HEADER_LIST_SIZE]});
+  }
+
+  if ((flags & (1 << IDX_SETTINGS_ENABLE_PUSH)) ==
+      (1 << IDX_SETTINGS_ENABLE_PUSH)) {
+    DEBUG_HTTP2("Setting enable push: %d\n",
+                buffer[IDX_SETTINGS_ENABLE_PUSH]);
+    entries.push_back({NGHTTP2_SETTINGS_ENABLE_PUSH,
+                       buffer[IDX_SETTINGS_ENABLE_PUSH]});
+  }
+
+  const size_t len = entries.size() * 6;
+  MaybeStackBuffer<char> buf(len);
+  ssize_t ret =
+      nghttp2_pack_settings_payload(
+        reinterpret_cast<uint8_t*>(*buf), len, &entries[0], entries.size());
+  if (ret >= 0) {
+    args.GetReturnValue().Set(
+      Buffer::Copy(env, *buf, len).ToLocalChecked());
```

Comment:
Either should work as the values should be identical. It would likely be worthwhile adding a `CHECK` just in case.